### PR TITLE
feat(derived_code_mappings): Support Javascript

### PR DESCRIPTION
**Pull request by armenzg**
This adds support for handling Javascript. More test cases are required.

Fixes WOR-2229

### CODE REVIEW

File: src/sentry/tasks/derive_code_mappings.py [hunk 1]
```diff
@@ -68,20 +69,23 @@ def derive_code_mappings(
...
+    if data["platform"] not in SUPPORTED_LANGUAGES:
         return []
+
     stacktraces = get_stacktrace(data)
     stacktrace_paths = set()
     for stacktrace in stacktraces:
         try:
-            paths = {frame["filename"] for frame in stacktrace["frames"]}
+            paths = {
+                frame["filename"] for frame in stacktrace["frames"] if frame.get("in_app") is True
```
Comment (armenzg) [inline src/sentry/tasks/derive_code_mappings.py:87]: This allows ignoring frames that the SDK has marked as a 3rd party module.
At the moment, the Python SDK marks all frames as in_app, thus, it was not necessary.

File: src/sentry/tasks/derive_code_mappings.py [hunk 2]
```diff
@@ -167,6 +172,9 @@ def set_project_codemappings(
                     "existing_code_mapping": cm,
                 },
             )
+        else:
+            stored_code_mappings.append(cm)
+    return stored_code_mappings
```
Comment (armenzg) [inline src/sentry/tasks/derive_code_mappings.py:177]: This is simply to enable the tests I've written.

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 3]
```diff
@@ -12,15 +14,69 @@
 from sentry.testutils.helpers.datetime import before_now, iso_format
 
 
-class TestIdentfiyStacktracePaths(TestCase):
+class BaseDeriveCodeMappings(TestCase):
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:17]: On A future PR, I will make the Python group of tests also inherit from this class.

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 4]
```diff
@@ -12,15 +14,69 @@
...
+                        "in_app": True,
+                    },
+                ]
+            },
+        }
+        self.event_data = self.store_event(data=self.test_data, project_id=self.project.id).data
+
+    def test_find_stacktrace_paths_single_project(self):
+        stacktrace_paths = identify_stacktrace_paths(self.event_data)
+        assert stacktrace_paths == ["./app/utils/handleXhrErrorResponse.tsx"]
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:51]: This shows that `../node_modules/@sentry/browser/node_modules/@sentry/core/esm/hub.js` will be ignored.

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 5]
```diff
@@ -12,15 +14,69 @@
...
         )
+        with patch(
+            "sentry.integrations.github.client.GitHubClientMixin.get_trees_for_org"
+        ) as mock_get_trees_for_org:
+            mock_get_trees_for_org.return_value = {
+                repo_name: RepoTree(
+                    Repo(repo_name, "master"), ["static/app/utils/handleXhrErrorResponse.tsx"]
+                )
+            }
+            code_mapping = derive_code_mappings(self.project.id, self.event_data)[0]
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:71]: This shows that a code mapping was generated.

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 6]
```diff
@@ -51,20 +107,18 @@ def setUp(self):
     def test_finds_stacktrace_paths_single_project(self):
         event = self.store_event(data=self.test_data, project_id=self.project.id)
 
-        with self.tasks():
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:54]: This and the following changes are unrelated. Just a refactor.

File: src/sentry/integrations/utils/code_mapping.py [hunk 7]
```diff
@@ -262,14 +265,9 @@ def _matches_current_code_mappings(self, src_file: str, frame_filename: FrameFil
...
 
-    def _potential_match(self, src_file: str, frame_filename: FrameFilename) -> bool:
-        """Tries to see if the stacktrace without the root matches the file from the
-        source code. Use existing code mappings to exclude some source files
-        """
-        # Exit early because we should not be processing source files for existing code maps
-        if self._matches_current_code_mappings(src_file, frame_filename):
-            return False
-
+    def _potential_match_with_transformation(
```
Comment (snigdhas) [inline src/sentry/integrations/utils/code_mapping.py:268]: should we add a docstring for the transformation being done?

File: src/sentry/integrations/utils/code_mapping.py [hunk 8]
```diff
@@ -285,6 +283,27 @@ def _potential_match(self, src_file: str, frame_filename: FrameFilename) -> bool
...
+        """
+        # Exit early because we should not be processing source files for existing code maps
+        if self._matches_current_code_mappings(src_file, frame_filename):
+            return False
+
+        match = False
+        if frame_filename.full_path.endswith(".py"):
+            match = self._potential_match_with_transformation(src_file, frame_filename)
+        else:
+            match = self._potential_match_no_transformation(src_file, frame_filename)
```
Comment (snigdhas) [inline src/sentry/integrations/utils/code_mapping.py:306]: No preference, but this could also be 

```suggestion
        if frame_filename.full_path.endswith(".py"):
            return self._potential_match_with_transformation(src_file, frame_filename)
        return self._potential_match_no_transformation(src_file, frame_filename)
```

File: src/sentry/integrations/utils/code_mapping.py [hunk 9]
```diff
@@ -285,6 +283,27 @@ def _potential_match(self, src_file: str, frame_filename: FrameFilename) -> bool
             match = (
                 split[0].rfind(f"/{frame_filename.root}") > -1 or split[0] == frame_filename.root
```
Comment (snigdhas) [inline src/sentry/integrations/utils/code_mapping.py:284]: IIUC, this is the main difference in the transformation - we might be able to clean up the way this is split at a later point.

File: src/sentry/tasks/derive_code_mappings.py [hunk 10]
```diff
@@ -68,20 +69,23 @@ def derive_code_mappings(
         report_project_codemappings(code_mappings, stacktrace_paths, project)
         return
 
-    set_project_codemappings(code_mappings, organization_integration, project)
+    return set_project_codemappings(code_mappings, organization_integration, project)
```
[the remark this answers is not in the document]
Comment (snigdhas) [reply src/sentry/tasks/derive_code_mappings.py:72]: We could also read it directly from the database with `RepositoryProjectPathConfig.objects.get`. Up to you on what you prefer

File: src/sentry/integrations/utils/code_mapping.py (hunk 8 above)
Comment (armenzg) [reply src/sentry/integrations/utils/code_mapping.py:336]: Yes both are possible but I follow this specific design pattern where returns are only at the end of a function. In larger functions it is a noticeably better pattern.

File: src/sentry/tasks/derive_code_mappings.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (armenzg) [reply src/sentry/tasks/derive_code_mappings.py:72]: That's a great idea! Thank you!

File: src/sentry/integrations/utils/code_mapping.py [hunk 11]
```diff
@@ -72,11 +72,23 @@ def filter_source_code_files(files: List[str]) -> List[str]:
 
 # XXX: Look at sentry.interfaces.stacktrace and maybe use that
 class FrameFilename:
-    def __init__(self, stacktrace_frame_file_path: str) -> None:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:75]: I'm shortening the variable name to `frame_file_path`.

File: src/sentry/integrations/utils/code_mapping.py [hunk 12]
```diff
@@ -72,11 +72,23 @@ def filter_source_code_files(files: List[str]) -> List[str]:
...
-        if stacktrace_frame_file_path.find("/") > -1:
-            # XXX: This code assumes that all stack trace frames are part of a module
-            self.root, self.file_and_dir_path = stacktrace_frame_file_path.split("/", 1)
+    def __init__(self, frame_file_path: str) -> None:
+        self.full_path = frame_file_path
+        self.extension = get_extension(frame_file_path)
+        if self.frame_type() == "packaged":
+            self._packaged_logic(frame_file_path)
+        else:
+            self._straight_path_logic(frame_file_path)
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:81]: These 5 lines are new.

File: src/sentry/integrations/utils/code_mapping.py [hunk 13]
```diff
@@ -72,11 +72,23 @@ def filter_source_code_files(files: List[str]) -> List[str]:
...
+
+    def frame_type(self) -> str:
+        type = "packaged"
+        if self.extension not in ["py"]:
+            type = "straight_path"
+        return type
+
+    def _packaged_logic(self, frame_file_path: str) -> None:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:89]: This logic is what was on `__init__`

File: src/sentry/integrations/utils/code_mapping.py [hunk 14]
```diff
@@ -88,8 +100,18 @@ def __init__(self, stacktrace_frame_file_path: str) -> None:
         else:
             self.root = ""
             self.dir_path = ""
-            self.file_and_dir_path = self.full_path
-            self.file_name = self.full_path
+            self.file_and_dir_path = frame_file_path
+            self.file_name = frame_file_path
+
+    def _straight_path_logic(self, frame_file_path: str) -> None:
+        # XXX: Write unit tests for these
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:107]: Most of the functionality is in this PR, however, I want more tests before we go live for EA.
We can merge this PR since the JS code would not yet be exercised.

File: src/sentry/integrations/utils/code_mapping.py [hunk 15]
```diff
@@ -210,18 +232,27 @@ def list_file_matches(self, frame_filename: FrameFilename) -> List[Dict[str, str
         return file_matches
 
     def _get_code_mapping_source_path(self, src_file: str, frame_filename: FrameFilename) -> str:
-        """Generate the source path of a code mapping
-        e.g. src/sentry/identity/oauth2.py (sentry/identity/oauth2.py) -> src/sentry
-        e.g. src/sentry/wsgi.py (sentry/wsgi.py) -> src/sentry
-        e.g. ssl.py -> raise NotImplementedError
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:216]: Move part of these comments in-line with the code.

File: src/sentry/integrations/utils/code_mapping.py [hunk 16]
```diff
@@ -210,18 +232,27 @@ def list_file_matches(self, frame_filename: FrameFilename) -> List[Dict[str, str
...
-        """
-        if frame_filename.dir_path != "":
-            source_path = src_file.rsplit(frame_filename.dir_path)[0].rstrip("/")
-            return f"{source_path}/"
-        elif frame_filename.root != "":
-            return src_file.rsplit(frame_filename.file_name)[0]
+        """Generate the source code root for a code mapping"""
+        source_code_root = None
+        if frame_filename.frame_type() == "packaged":
+            if frame_filename.dir_path != "":
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:238]: This block has been right shifted. If you use "Hide whitespace" it is easier to see any differences.
image

File: src/sentry/integrations/utils/code_mapping.py [hunk 17]
```diff
@@ -210,18 +232,27 @@ def list_file_matches(self, frame_filename: FrameFilename) -> List[Dict[str, str
...
+            else:
+                # ssl.py -> raise NotImplementedError
+                raise NotImplementedError("We do not support top level files.")
         else:
-            raise NotImplementedError("We do not support top level files.")
+            # static/app/foo.tsx (./app/foo.tsx) -> static/app/
+            # static/app/foo.tsx (app/foo.tsx) -> static/app/
+            source_code_root = src_file.replace(
+                frame_filename.file_and_dir_path, frame_filename.root.replace("./", "")
+            )
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py]: This is the new block for JS support.

File: src/sentry/integrations/utils/code_mapping.py [hunk 18]
```diff
@@ -262,14 +293,14 @@ def _matches_current_code_mappings(self, src_file: str, frame_filename: FrameFil
...
-        if self._matches_current_code_mappings(src_file, frame_filename):
-            return False
+    def _potential_match_with_transformation(
+        self, src_file: str, frame_filename: FrameFilename
+    ) -> bool:
+        """Determine if the frame filename represents a source code file.
 
+        Languages like Python include the package name at the front of the frame_filename, thus, we need
+        to drop it before we try to match it.
+        """
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:303]: This is the doc string you requested.

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 19]
```diff
@@ -12,15 +14,71 @@
...
+            mock_get_trees_for_org.return_value = {
+                repo_name: RepoTree(
+                    Repo(repo_name, "master"), ["static/app/utils/handleXhrErrorResponse.tsx"]
+                )
+            }
+            derive_code_mappings(self.project.id, self.event_data)
+            code_mapping = RepositoryProjectPathConfig.objects.all()[0]
+            # ./app/foo.tsx -> foo.tsx -> static/app/foo.tsx
+            assert code_mapping.stack_root == "./app/"
+            assert code_mapping.source_root == "static/app/"
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:75]: On Sentry, this value is `.` while the source_root is `static`. I consider `.` in this case the same as the blank stack root which I disfavour because it mostly matches everything.

**Review by snigdhas — APPROVED** [review]
